Frame simulator

skyfrench opened the pull request:
Initial implementation of a 'frame simulator' base plugin class / application for consideration; pull request also opened in dls-controls/excalibur-detector with example implementation.

ulrikpedersen commented:
ah, it looks like there are some commits from the 'dls-master' branch in this PR. Those commits are DLS-specific and shouldn't get merged in here (I think). 

@skyfrench: I guess that's what you said yesterday about having to figure out what branch to merge to/from...? I'm not completely sure what is the easiest way to resolve this. Can you rebase your branch with your main changeset (i.e. those latest 7 commits) on top of odin-detector/master?

ulrikpedersen commented:
The travis build seems to fail due to cmake not finding pcap. Try adding package "libpcap-dev" to the list in .travis.yml

MattTaylorDLS commented on frameSimulator/src/FrameSimulatorPlugin.cpp:
```diff
@@ -0,0 +1,53 @@
+     * \param[out] true (false) if required program options are (not) specified
+     * and the simulator plugin is not ready to use
+     *
+     * Drived plugins must additionally call this base method even if overridden
```
Typo: Derived

MattTaylorDLS commented on frameSimulator/src/pcapFrameSimulatorPlugin.cpp:
```diff
@@ -0,0 +1,157 @@
+        return sendto(m_socket, packet.data, packet.size, 0, (struct sockaddr *) (&m_addrs[curr_port_index]), sizeof(m_addrs[curr_port_index]));
+    }
+
+    void pcapFrameSimulatorPlugin::populate_options(po::options_description& config) {
```
Needs function comment

MattTaylorDLS commented on frameSimulator/src/pcapFrameSimulatorPlugin.cpp:
```diff
@@ -0,0 +1,157 @@
+
+    }
+
+    pcapFrameSimulatorPlugin::~pcapFrameSimulatorPlugin() {
```
Needs function comment

MattTaylorDLS commented on frameSimulator/src/pcapFrameSimulatorPlugin.cpp:
```diff
@@ -0,0 +1,157 @@
+        replayer->prepare_packets(hdr, buffer);
+    }
+
+    void pcapFrameSimulatorPlugin::simulate() {
```
Needs function comment

MattTaylorDLS commented on frameSimulator/src/pcapFrameSimulatorPlugin.cpp:
```diff
@@ -0,0 +1,157 @@
+
+    }
+
+    /** Packet callback function
```
Needs arguments in function comment

MattTaylorDLS commented on frameSimulator/src/FrameSimulatorApp.cpp:
```diff
@@ -0,0 +1,206 @@
+static const FrameSimulatorOption<std::string> opt_logconfig("logconfig", "Set the log4cxx logging configuration file");
+
+/** Load requested plugin
+ * /param[in] map of command line specified variables
```
I think the param comment should have the name of the variable (e.g. "vm"). [This comment also applies to other instances where the variable is not named in the comment in this PR]

MattTaylorDLS commented on frameSimulator/src/FrameSimulatorApp.cpp:
```diff
@@ -0,0 +1,206 @@
+    boost::filesystem::path libraryPathAndName = boost::filesystem::path(opt_libpath.get_val(vm)) /
+        boost::filesystem::path("lib" + pluginClass + ".so");
+
+boost::shared_ptr<FrameSimulator::FrameSimulatorPlugin> plugin = \
```
indentation

MattTaylorDLS commented on frameSimulator/src/FrameSimulatorApp.cpp:
```diff
@@ -0,0 +1,206 @@
+
+}
+
+static bool has_suffix(const std::string &str, const std::string &suffix)
```
Needs function comment

MattTaylorDLS commented on frameSimulator/src/FrameSimulatorApp.cpp:
```diff
@@ -0,0 +1,206 @@
+           str.compare(str.size() - suffix.size(), suffix.size(), suffix) == 0;
+}
+
+int parse_arguments(int argc, char** argv, po::variables_map& vm, LoggerPtr& logger,
```
Needs function comment

MattTaylorDLS commented on frameSimulator/src/FrameSimulatorApp.cpp:
```diff
@@ -0,0 +1,206 @@
+
+            plugin = get_requested_plugin(vm);
+
+            if (plugin)
```
Think this should have { } even though it's only one line within the if.

MattTaylorDLS commented on frameSimulator/src/FrameSimulatorApp.cpp:
```diff
@@ -0,0 +1,206 @@
+        po::variables_map vm;
+        parse_arguments(argc, argv, vm, logger, plugin);
+
+        if (!plugin)
```
Think this should have { }

MattTaylorDLS commented on frameSimulator/src/FrameSimulatorApp.cpp:
```diff
@@ -0,0 +1,206 @@
+            LOG4CXX_ERROR(logger, "Unable to create simulator plugin, application will terminate");
+
+        // Setup plugin from command line arguments
+        if (plugin->setup(vm))
```
{ }

MattTaylorDLS commented on frameSimulator/src/FrameSimulatorOption.cpp:
```diff
@@ -0,0 +1,21 @@
+      void set_list_option(const std::string& option_val, std::vector<std::string>& list) {
+          boost::split(list, option_val, boost::is_any_of(","), boost::token_compress_on);
+      }
+      void set_optionallist_option(const std::string& option_val, boost::optional<std::vector<std::string> >& list) {
```
Should have separate function comments for each function

MattTaylorDLS commented on frameSimulator/src/FrameSimulatorPlugin.cpp:
```diff
@@ -0,0 +1,53 @@
+    /** Setup frame simulator plugin class from store of command line options
+     *
+     * \param[in] store of given command line options
+     * \param[out] true (false) if required program options are (not) specified
```
Should be "return" not "param"

MattTaylorDLS commented on frameSimulator/src/pcapFrameSimulatorPlugin.cpp:
```diff
@@ -0,0 +1,157 @@
+
+namespace FrameSimulator {
+
+    pcapFrameSimulatorPlugin::pcapFrameSimulatorPlugin() : FrameSimulatorPlugin() {
```
Needs function comment

MattTaylorDLS commented on frameSimulator/src/pcapFrameSimulatorPlugin.cpp:
```diff
@@ -0,0 +1,157 @@
+
+    }
+
+    bool pcapFrameSimulatorPlugin::setup(const po::variables_map& vm) {
```
Needs function comment

MattTaylorDLS commented on frameSimulator/src/pcapFrameSimulatorPlugin.cpp:
```diff
@@ -0,0 +1,157 @@
+        opt_packetgap.get_val(vm, packet_gap);
+        opt_dropfrac.get_val(vm, drop_frac);
+
+        if (opt_droppackets.is_specified(vm))
```
{ }

MattTaylorDLS commented on cmake/FindPCAP.cmake:
```diff
@@ -0,0 +1,47 @@
+    PCAP_ROOT_DIR
+    PCAP_INCLUDE_DIR
+    PCAP_LIBRARY
+)
```
No new line at end of file

MattTaylorDLS commented on frameSimulator/include/FrameSimulatorOptions.h:
```diff
@@ -0,0 +1,18 @@
+
+}
+
+#endif
```
No new line at end of file

MattTaylorDLS commented on frameSimulator/include/FrameSimulatorPlugin.h:
```diff
@@ -0,0 +1,61 @@
+
+}
+
+#endif
```
No new line at end of file

MattTaylorDLS commented on frameSimulator/src/FrameSimulatorApp.cpp:
```diff
@@ -0,0 +1,206 @@
+
+    return 0;
+
+}
```
No new line at end of file

MattTaylorDLS commented on frameSimulator/src/FrameSimulatorOption.cpp:
```diff
@@ -0,0 +1,21 @@
+          list = tmp_list;
+      }
+
+}
```
No new line at end of file

MattTaylorDLS commented on frameSimulator/src/FrameSimulatorPlugin.cpp:
```diff
@@ -0,0 +1,53 @@
+
+    }
+
+}
```
No new line at end of file

MattTaylorDLS commented on frameSimulator/src/Packet.cpp:
```diff
@@ -0,0 +1,6 @@
+namespace FrameSimulator {
+
+
+}
```
No new line at end of file

MattTaylorDLS commented on frameSimulator/src/pcapFrameSimulatorPlugin.cpp:
```diff
@@ -0,0 +1,157 @@
+
+    }
+
+}
```
No new line at end of file

ajgdls commented on frameSimulator/src/FrameSimulatorOption.cpp:
```diff
@@ -0,0 +1,26 @@
+     * /param[in] string of comma separated values
+     * /param[out] required output list as std::vector
+     */
+    void set_optionallist_option(const std::string &option_val, boost::optional <std::vector<std::string> > &list) {
```
To be able to compile this using DLS build program on RHEL6 I had to add a space between the two >> brackets.

ajgdls replied:
Whoops @skyfrench I see you have already committed this change just now, and so this isn't some problem with compiler versions RHEL6/7.

timcnicholls commented:
Also noticed that the FindPCAP.cmake file doesn't announce itself like the ZeroMQ, log4cxx, HDF ones do:
```
$ cmake ..
-- Boost version: 1.67.0
-- Found the following Boost libraries:
--   program_options
--   system
--   filesystem
--   unit_test_framework
--   date_time
--   thread

Looking for log4cxx headers and libraries
-- Root dir: /aeg_sw/tools/CentOS7-x86_64/log4cxx/0-10-0/prefix
-- Include directories: /aeg_sw/tools/CentOS7-x86_64/log4cxx/0-10-0/prefix/include/log4cxx
-- Libraries: /aeg_sw/tools/CentOS7-x86_64/log4cxx/0-10-0/prefix/lib/liblog4cxx.so

Looking for ZeroMQ headers and libraries
-- Root dir: /aeg_sw/tools/CentOS7-x86_64/zeromq/4-2-3/prefix
-- Include directories: /usr/include
-- Libraries: /lib64/libzmq.so
-- Found PCAP: /usr/lib64/libpcap.so (Required is at least version "1.4.0") 
-- Performing Test PCAP_LINKS
-- Performing Test PCAP_LINKS - Success

Searching for HDF5
-- HDF5_ROOT set: /aeg_sw/tools/CentOS7-x86_64/hdf5/1-10-4/prefix
Determining odin-data version
-- Git describe version: 0.6.1-22-g05581ef
-- major:0 minor:6 patch:1 sha1:g05581ef
-- short version: 0.6.1
-- HDF5 include files:  /aeg_sw/tools/CentOS7-x86_64/hdf5/1-10-4/prefix/include
-- HDF5 libs:           /aeg_sw/tools/CentOS7-x86_64/hdf5/1-10-4/prefix/lib/libhdf5.so/aeg_sw/tools/CentOS7-x86_64/hdf5/1-10-4/prefix/lib/libhdf5_hl.so
-- HDF5 defs:           
-- Configuring done
-- Generating done
-- Build files have been written to: /u/tcn45/develop/projects/excalibur/odin-data/build
```
which makes it look like part of the ZeroMQ resolution step. Would it be possible to add a `Looking for ...` message call?

skyfrench commented:
@timcnicholls suggested changes made :)

skyfrench commented:
I've added --version to frameProcessor (and removed the version text from the help message) for consistency. The version reporting now matches for all three.

timcnicholls commented:
Fantastic, thanks!
